### ISSUE TITLE
Making Chromium more secure

### ISSUE THREAD

**Issue by boredsquirrel** [secureblue/secureblue#161]
I am not sure if this is already done and I just dont get the changes, but:

- HTTPS only mode by default
- disabling the password manager?
- disable preloading of websites?
- not sure about safe browsing, would keep it disabled (by default it sends "malicious websites" to Google, whatever that is)

I think it would be great if Ublock origin could be preinstalled without any problems. On Chromium this either requires a Google Account, or the [chromium-webstore](https://github.com/NeverDecaf/chromium-web-store/issues) Addon

I would not consider this a privacy setting, because the alternative would be requiring a Google Account to get Addon Updates.

Also adding these to Ublock origins "custom filters":

```
*.zip
*.mov
```

as they should be opt-in with a warning that it may be malware disguised as a file

**Comment by RoyalOughtness** [comment]
Thanks for your suggestions:

> HTTPS only mode by default

This is a good idea. Unfortunately I'm not sure if there's a command line switch for it though. According to [this list](https://peter.sh/experiments/chromium-command-line-switches/), there isn't one. I just added this to the postinstall readme and we can put other security-relevant chromium config recommendations there as well for things that don't have a command line flag: https://github.com/secureblue/secureblue/blob/live/POSTINSTALL-README.md#chromium-settings

> disabling the password manager?

While I agree that the password manager is annoying, there's no security relevance to changing the default :smile:

> disable preloading of websites?

What's the security relevance of this?

> not sure about safe browsing, would keep it disabled (by default it sends "malicious websites" to Google, whatever that is)

This should be enabled as it's a security feature.

> I think it would be great if Ublock origin could be preinstalled without any problems. 

Couple things. One is that MV3 should be preferred due to the [security advantages over MV2](https://developer.chrome.com/docs/extensions/develop/migrate/improve-security), so UBO-Lite should be recommended.

Second, unfortunately preinstalling extensions is cumbersome at best. Any method needed is going to involve some kind of user input anyways, so given how simple it is to install from the store, I've put it in the [POSTINSTALL-README](https://github.com/secureblue/secureblue/blob/live/POSTINSTALL-README.md) instead

**Comment by boredsquirrel** [comment]
> not sure if there's a command line switch

yeah I find Chromium very hard to configure, compared to Firefox where you would just put everything in a user.js (this is not systemwide though and I also have no idea of systemwide configs on Linux). On Windows there are group policies, but that is completely Windows specific.

I will try to debloat & harden Brave like this, but it is really hard to even find any info on how to set those settings from a file, no idea.

The password manager is completely offline on Chromium, and GrapheneOS trusts the Vanadium one, so agree.

> What's the security relevance of this?

Preloading: Afaik the Browser opens random links that you ***may*** click, which is a pretty weird concept and in some scenarios, for example an Adblocker not blocking everything, may lead to bad code executed without any form of user control.

> MV3 should be preferred

Yes, mv3 is probably a security advantage, just not for adblockers. It prevents addons from loading config files from the web that is not tunneled through Google. This means Addons may get less updates, because Google slows the "control procedure" down on purpose, with the sole reason to be faster with their ads than the Adblocker can update.

This means Google fighting with ubo lite could mean many other filters would also be delayed, as they all come bundled in one addon update.

https://github.com/gorhill/uBlock/wiki/uBlock-Origin-works-best-on-Firefox

UBO already needed to avoid using Webassembly, to prevent being blocked by the Chrome Web Store. This indicates they may comply with their permissive adblocking to still get security updates shipped, but that will result in way worse adblocking performance.

Not opening a Firefox vs. Chromium debate here, but Chromium simply doesnt support Adblock and even Malware protection without Google in this configuration. This is not "not adding random privacy features", this is "forcing people to use Google to get basic security features". Both with safe browsing and with installing addons.

The situation is bad and it looks like Chromium will need to have robust support for MV3 addon sideloading, which the "chromium webstore" addon supposedly supports. The other option is not only forcing people to use Google for addon downloads (and thus somewhat tied to the browser), but also uploading "malicious" domains to Google, whatever many false positives would be among those.

For security the best way would be to use safe browsing in "strict" mode and use UBO lite and hope this will have some usability improvements. But this is of course not privacy friendly at all, but actively anti-privacy compared to the current state.

I know this is out of scope of this project though, and it will end up in other projects based on secureblue, if not here.

**Comment by RoyalOughtness** [comment]
> Preloading: Afaik the Browser opens random links that you may click, which is a pretty weird concept and in some scenarios, for example an Adblocker not blocking everything, may lead to bad code executed without any form of user control.

I see. I just checked though and it looks like the chromium rpm already has this off by default, so we should be good.

> just not for adblockers. 

Nope, the same security advantages apply for adblockers. Functionality issues don't negate security improvements. They're a consideration, but not a negation.

> It prevents addons from loading config files from the web that is not tunneled through Google

It prevents them from loading arbitrary content, yes. This is a good thing from a security perspective.

> This means Addons may get less updates, because Google slows the "control procedure" down on purpose, with the sole reason to be faster with their ads than the Adblocker can update.

That's a negative side effect of one of the security improvements, yes. Whether it's intentional or not is frankly not terribly relevant.

> UBO already needed to avoid using Webassembly, to prevent being blocked by the Chrome Web Store. 

That's great :smile:

> this is "forcing people to use Google to get basic security features". 

No one is being forced to do anything. If anything, it's firefox's lack of security parity that "forces" people to use chromium to get proper security. The answer here is for firefox to catch up. Until then, chromium based browsers are the only recommendable browsers from a security perspective. And on top of all of that, MV3 is a security improvement and getting rid of MV2 is a security improvement.

> the other option is not only forcing people to use Google for addon downloads 

Again, it's not a matter of force. Also, downloading addons through a distribution mechanism that provides signing and verification is preferable. Much like the most secure way to install applications on GrapheneOS is to use the Google Play Store.

> I know this is out of scope of this project though

Yep. And even if you do care strongly about privacy, sacrificing security for the sake of privacy isn't worth it. But like you said this is all offtopic :smile:

**Comment by wandapeter** [comment]
Just FYI

Configuring chrome is possible via config files. Check out this STIG
https://static.open-scap.org/ssg-guides/ssg-chromium-guide-index.html

Or this very extensive example configuration by Divest Groups "Brace" project:
https://github.com/divestedcg/Brace/blob/master/brace/etc/chromium/policies/managed/brace.json

This is the official administrators documentation on config files:
https://www.chromium.org/administrators/linux-quick-start/

**Comment by boredsquirrel** [comment]
> I just checked though and it looks like the chromium rpm already has this off by default, so we should be good.

I am confused, because on my install it was not disabled by default. Is the GUI showing stuff that is not actually true, like there being dominant configs that override it?

Thanks @34N0 this is very useful! So this would allow settings like HTTPS-only?

> It prevents them from loading arbitrary content, yes. This is a good thing from a security perspective.

I think this discussion is not leading anywhere because the current issues that you mentioned, there simply is no good alternative. But this means, should private browsing be preset to be the most "secure" variant then? I think this is very invasive on privacy. Just FYI GrapheneOS has it disabled by default, but I dont know what they use instead.

If there would be any alternative solution to this problem, it would be really great.

**Comment by wandapeter** [comment]
yes i think setting https only mode would look like:
`/etc/chromium/policies/managed/secureblue.json (755)`:
```json
{
  "HttpsOnlyMode": "force_enabled",
}
```

**Comment by RoyalOughtness** [comment]
> I am confused, because on my install it was not disabled by default. Is the GUI showing stuff that is not actually true, like there being dominant configs that override it?

That's strange, we should set it in config then if possible.

> I think this is very invasive on privacy. Just FYI GrapheneOS has it disabled by default, but I dont know what they use instead.

Has what disabled by default? the password manager? you replied to my comment about MV3 so this is confusing.

> If there would be any alternative solution to this problem, it would be really great.

To which problem?

**Comment by RoyalOughtness** [comment]
reopening this as we can use the mechanism @34N0 mentioned to set some additional config

**Comment by boredsquirrel** [comment]
@qoijjj 

> Has what disabled by default? the password manager? you replied to my comment about MV3 so this is confusing.

"should ~~private~~ safe browsing be preset to be the most "secure" variant then? I think this is very invasive on privacy. Just FYI GrapheneOS has it disabled by default" 

GrapheneOS has "safe browsing" disabled, but I dont know how they deal with the consequences.

> To which problem?

The problem relying on Google (directly, not proxied and anonymized like in Firefox and afaik Brave) for security. Not only safe browsing but also the problem that UBO-lite updates may be delayed.

**Comment by boredsquirrel** [comment]
@34N0 The DivestedCG Policy is really cool. Looking through it currently.

I think many settings could be used, but discussed which of those. 

For example you can force preinstall addons, I would suggest UBO-lite and NoScript, which can both be disabled or configured to do nothing.

Also you can allowlist ("whitelist") addons, I think this is great but will need some maintenance. Still I think it should be used. All addons should be tested, known and "recommended" (following chromium guidelines". I am searching for commonly used addons currently:

```
- Plasma-Integration: cimiefiiaegbelhefglklhhakcgmhkai
- GNOME-Integration: gphhapmejobijbbhgpjhcjognlahblep
- UBlock Origin Lite: ddkjiahejlhfcafbddmgiahcphecmpfh
- NoScript: doojmbjmlfjjnbmnoijecmcbfeoakpjm
- UBlock Origin (full variant): cjpalhdlnbpafiamejdnhcphjbkeiagm
- KeepassXC-Browser: oboonakemofpalcgghocfoadofidjkkk
- Bitwarden: nngceckbapebfimnlniiiahkandclblb
- ProtonPass: ghmbeldphafepmbegfdlkpapadhbakde

More:
- Snowflake: mafpmfcccpbjnhfhjnllmmalhifmlcie
- SingleFile: mpiodijhokgodhhofbcjdecpffjipkle
- DownloadAllImages: nnffbdeachhbpfapjklmpnmjcgamcdmm
- Privacy Redirect: pmcmeagblkinmogikoikkdjiligflglb (UNMAINTAINED)
- Youtube to Invidious Switcher: gebpglgiomkfhcipdikkbmmjggcnonpa
- Undo Closed Tab Button: ieehkmoiljghfkejgahoheemdjpdinml
- Find Sci Paper: ocofgmnfmjndinnmdimpmijogpaljmal
- Dark Reader: eimadpbcbfnmbkopoojfekhnkhdbieeh ### dark background & light text from firefox is missing, but that one is well known and also follows Chromium Guidelines
- Offline QR-Code Generator: fehmldbcmhbdkofkiaedfejkalnidchm
- Offline QR Code Generator/Editor: kfhbhjigpkcbpmknfomdobahejfajado
- Zotero Connector: ekhagklcjbdpajgpjgmbionohlpdbjgc
- Mailvelope: kajibbejlbohfaggdiogboambcijhkke
- - Privacy Badger: pkehgijcmpdhfbdbbnkijodmdjhbjlgp
- Cookie Autodelete: fhcgjolkccmbidfldomjliifgaodjagh
- ReadBee: phjbepamfhjgjdgmbhmfflhnlohldchb
- Readline: lgfibgggkoedaaihmmcifkmdfdjenlpp
- OneTab: chphlpgkkbolifaimnlloiipkdnihall
- Vimium: dbepggeogbaibhgnhhndojpepiihcmeb
- Vimium C: hfjbmagddngcpeloejdejnfgbamkjaeg
- PDF Viewer for Vimium C: nacjakoppgmdcpemlfnfidegmlhipddanj

DivestedCG:
- enhanced-h264ify: omkfmpieigblcllmkgbflkikinpkodlk
```

it would probably be smarter to just preinstall addons and maybe block others, not maintain an allowlist. At the same time I can imagine there are malicious addons out there, and the Browser is often the biggest attack surface. Installation of shady addons could be an attack vector. And we could create a Github issue template to request an addon to be added like this

## Request an Extension to be allowlisted in Chromium

- [ ] I have checked for duplicate issues
- [ ] The Addon has a "recommended" tag
- [ ] I have considered unblocked alternatives and they don't match
- [ ] The Addon is known and has a good privacy policy

What Extension do you want allowlisted (include its full ID) ?
``` text```

Reasoning to request it:
``` text```

**Comment by RoyalOughtness** [comment]
https://github.com/secureblue/secureblue/commit/2f416b62b3703c73329273cbe2f9b0bf3931c33e

**Comment by boredsquirrel** [comment]
what about the allowlist for addons? This means some good amount of work and dealing with users wanting their favourite Addon to be added. But in my experience especially the Chrome Web Store is full of garbage, and currently I am not sure if it would even be prohibited to sideload addons.

I would like to include chromium-webstore as a possibility to sideload.

**Comment by RoyalOughtness** [comment]
> what about the allowlist for addons? This means some good amount of work and dealing with users wanting their favourite Addon to be added. But in my experience especially the Chrome Web Store is full of garbage, and currently I am not sure if it would even be prohibited to sideload addons.

I don't really understand the question. The extension allowlist is for corporations who want to control what their employees can install. This has nothing to do with secureblue's scope.

> I would like to include chromium-webstore as a possibility to sideload.

Include it where? If you mean install it by default, definitely not as it would explicitly be a security degradation. The most secure way to install extensions is via the chrome webstore, much like the most secure way to install applications on grapheneos is via google play.

**Comment by boredsquirrel** [comment]
agree an allowlist may be too strict.

on Chromium (not ungoogled-chromium) you can install "Chromium Webstore" only by enabling developer mode and drag&dropping the manually (e.g. via `wget`) doenloaded Chromium.Web.Store.crx file.

Downloading the file does not work because it is external and thus blocked. This could be fixed to allow users to download Addons from the Chrome Webstore without an Account.

```
	"ExtensionInstallAllowlist": "ocaahdebbfolfmndjeplogmgcagdmblk",
```

adding this to the policy ~~will~~ only allow this exact addon to be sideloaded. Would be a great addition imo, I will test it first.

Edit: weird didnt work, but enabling developer mode and installing it again keeps the same ID. I think it would be preferred to allow installing this one addon, without developer mode.

Would it be wanted to document the installation of the addon? Its primary purpose is to avoid needing a Google account associated with the browser.

**Comment by RoyalOughtness** [comment]
> Edit: weird didnt work, but enabling developer mode and installing it again keeps the same ID. I think it would be preferred to allow installing this one addon, without developer mode.

If _you_ want to use it, you still can as you mentioned. Please stop bringing this up, it's yet another example of something specific to your use case that's a security anti-pattern. 

> Would it be wanted to document the installation of the addon? Its primary purpose is to avoid needing a Google account associated with the browser.

No, as I already said above. Using this extension is a security degradation. Also, you don't need a google account to use the chrome webstore, so that part is just factually wrong.

**Comment by boredsquirrel** [comment]
Thanks, forgot to check and maybe overlooked it, that you dont need a Google Account.

The idea was to avoid (and having the possibility to completely lock) the "extension developer mode", which allows the installation of arbitrary addons without any signing key.

In that case it wouldn't have been a specific use case but strictly usability focused. To my knowledge you couldnt download addons on Chromium without an account, not sure when they changed that.

✅

So, continuing that thought, maybe we should think about disabling the "extension developer mode" which is not currently done.
